src/torchmetrics/detection/iou.py

```diff
@@ -213,20 +216,36 @@ def _get_gt_classes(self) -> list:
 
     def compute(self) -> dict:
         """Computes IoU based on inputs passed in to ``update`` previously."""
-        score = torch.cat([mat[mat != self._invalid_val] for mat in self.iou_matrix], 0).mean()
+        # compute global IoU score using only valid values.
+        valid_matrices = [
+            mat[mat != self._invalid_val] for mat in self.iou_matrix if torch.any(mat != self._invalid_val)
+        ]
+        score = torch.cat(valid_matrices, 0).mean() if valid_matrices else torch.tensor(0.0)
```

Comment:
```suggestion
        score = torch.cat(valid_matrices, 0).mean() if valid_matrices else torch.tensor(0.0, device=self.device)
```
gpu tests are failing due to difference in device later in the code